// Copyright (c) 2023 dingodb.com, Inc. All Rights Reserved
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

import "common.proto";

package dingodb.pb.meta;

option java_package = "io.dingodb.meta";
option cc_generic_services = true;

// message MetricStat {
//   double disk_used_percent = 1;
//   double mem_used_percent = 2;
//   double cpu_used_percent = 3;
//   double disk_io_percent = 4;
// }

// message InstanceInfo {
//   dingodb.pb.common.Location address = 1;
//   MetricStat stat = 3;
// }

// message AllocateStoreIdRequest {
//   dingodb.pb.common.Location address = 1;
//   string resourceTag = 2;
//   int64 startTime = 3;
// }

// message AllocateStoreIdRes {
//   string storeId = 1;
// }

message GetSchemaRequest {
  dingodb.pb.common.CommonId schema_id = 1;
}

message GetSchemaResponse {
  repeated dingodb.pb.common.Schema schemas = 1;
}

message GetTablesRequest {
  repeated dingodb.pb.common.CommonId schema_id = 1;
}

message GetTablesResponse {
  repeated dingodb.pb.common.Table tables = 1;
}

message CreateTableRequest {
  dingodb.pb.common.TableDefinition table_definition;
}

message CreateTableResponse {
  repeated dingodb.pb.common.Table table = 1;
}

message DropTableRequest {
  dingodb.pb.common.CommonId table_id = 1;
}

message DropTableResponse {}

// message GetTablePartitionsReq {
//   dingodb.pb.common.CommonId table_id = 1;
// }

// message GetTablePartitionsRes {
//   repeated dingodb.pb.common.TablePartition table_partitions = 1;
// }

// message GetRegionsReq {
//   dingodb.pb.common.CommonId partition_id = 1;
// }

// message GetRegionsResponse {
//   dingodb.pb.common.TableRegions table_regions = 1;
// }

service MetaService {
  // query schema/table/partition/regions
  rpc listSchemas(GetSchemaRequest) returns (GetSchemaResponse);
  rpc listTables(GetTablesRequest) returns (GetTablesResponse);

  // rpc listPartitions(GetTablePartitionsReq) returns (GetTablePartitionsRes);
  // rpc listRegions(GetRegionsReq) returns (GetRegionsResponse);

  // Table DDL
  rpc createTable(CreateTableRequest) returns (CreateTableResponse);
  rpc dropTable(DropTableRequest) returns (DropTableResponse);

  // todo
  // Operation about User, such as create/drop/grant.

  // Schema, CreateSchema/DropSchema/ getAllTableFromSchema / getSubSchema

  // Operation about Metrics
}
